### PR TITLE
[WIP] Separate Self-Aware

### PR DESCRIPTION
**Pull request by paxcema**
Very early implementation that separates the self-aware network from the main NN into its own module to have a training loop where each one has its separate optimizer. See #217. Testing and benchmarking is still pending.

### CODE REVIEW

File: lightwood/mixers/helpers/selfaware.py [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+
+        for ind in range(len(awareness_net_shape) - 1):
+            rectifier = torch.nn.SELU
+            awareness_layers.append(torch.nn.Linear(awareness_net_shape[ind], awareness_net_shape[ind + 1]))
+            if ind < len(awareness_net_shape) - 2:
+                awareness_layers.append(rectifier())
+
+        self.net = torch.nn.Sequential(*awareness_layers)
+
+        for layer in self.net:
```
Comment (paxcema) [inline lightwood/mixers/helpers/selfaware.py:30]: Move to a helper that is shared

File: lightwood/mixers/nn/nn.py [hunk 2]
```diff
@@ -83,7 +85,9 @@ def build_confidence_normalization_data(self, ds, subset_id=None):
 
             with torch.no_grad():
                 if self.is_selfaware:
-                    outputs, awareness = self.net(inputs)
+                    outputs = self.net(inputs)
+                    aware_in = torch.cat((inputs, outputs), 1)
```
Comment (paxcema) [inline lightwood/mixers/nn/nn.py:89]: Add cat into selfaware forward

File: lightwood/mixers/nn/nn.py [hunk 3]
```diff
@@ -320,7 +324,9 @@ def predict(self, when_data_source, include_extra_data=False):
 
             with torch.no_grad():
                 if self.is_selfaware:
-                    output, awareness = self.net(inputs)
+                    output = self.net(inputs)
```
Comment (paxcema) [inline lightwood/mixers/nn/nn.py:327]: set eval mode

File: lightwood/mixers/nn/nn.py (hunk 2 above)
Comment (paxcema) [reply lightwood/mixers/nn/nn.py:89]: add eval

File: lightwood/mixers/nn/nn.py [hunk 4]
```diff
@@ -619,13 +621,16 @@ def iter_fit(self, ds, initialize=True, subset_id=None):
...
+                self.selfaware_optimizer.zero_grad()
 
                 # forward + backward + optimize
-                # outputs = self.net(inputs)
+                outputs = self.net(inputs)
                 if self.is_selfaware:
-                    outputs, awareness = self.net(inputs)
-                else:
-                    outputs = self.net(inputs)
+                    if self.start_selfaware_training:
```
Comment (paxcema) [inline lightwood/mixers/nn/nn.py:629]: Flags: selfaware_training_standalone, _with_net.

File: lightwood/mixers/nn/nn.py [hunk 5]
```diff
@@ -658,14 +662,11 @@ def iter_fit(self, ds, initialize=True, subset_id=None):
...
 
-
                 if CONFIG.MONITORING['batch_loss']:
                     self.monitor.plot_loss(loss.item(), self.total_iterations, 'Targets Batch Loss')
 
-
-
                 if awareness_loss is not None:
-                    awareness_loss.backward(retain_graph=True)
+                    awareness_loss.backward()
```
Comment (paxcema) [inline lightwood/mixers/nn/nn.py:669]: if training_encoder: retain_graph

File: lightwood/mixers/helpers/selfaware.py (hunk 1 above)
Comment (George3d6) [reply lightwood/mixers/helpers/selfaware.py:30]: Ideally keep them separate actually, since we might want to differentiate between the deafultNet and the selfaware net, we can move boilerplate to helper functions as we go.

File: lightwood/mixers/nn/nn.py [hunk 6]
```diff
@@ -320,7 +325,9 @@ def predict(self, when_data_source, include_extra_data=False):
 
             with torch.no_grad():
                 if self.is_selfaware:
-                    output, awareness = self.net(inputs)
+                    self.selfaware_net.eval()
```
Comment (George3d6) [inline lightwood/mixers/nn/nn.py:328]: it seems like the main network is not set to `eval`, if so we should also do this here. If it is set, but somewhere else, we should set the selfaware to eval there as well.

File: lightwood/mixers/nn/nn.py [hunk 7]
```diff
@@ -83,7 +86,9 @@ def build_confidence_normalization_data(self, ds, subset_id=None):
 
             with torch.no_grad():
                 if self.is_selfaware:
-                    outputs, awareness = self.net(inputs)
+                    self.selfaware_net.eval()
```
Comment (George3d6) [inline lightwood/mixers/nn/nn.py:89]: it seems like the main network is not set to `eval`, if so we should also do this here. If it is set, but somewhere else, we should set the selfaware to eval there as well.

File: lightwood/mixers/helpers/selfaware.py [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+        output = self.net(aware_in)
+        return output
+
+
+
+if __name__ == '__main__':
+    print("Various tests that need to be aggregated in tests/ before PRing")
+    sa = SelfAware(100, 50, 1)
+
+    print(sa)
```
Comment (George3d6) [inline lightwood/mixers/helpers/selfaware.py:63]: Remove or move under the current test structure.

File: lightwood/mixers/nn/nn.py (hunk 7 above)
Comment (paxcema) [reply lightwood/mixers/nn/nn.py:89]: It is set at the start of this function, so as it stands it seems fine to call .eval() for the self-aware here.

File: lightwood/mixers/nn/nn.py (hunk 6 above)
Comment (paxcema) [reply lightwood/mixers/nn/nn.py:328]: Same as in the previous comment
